src/Middleware/HeaderPropagation/src/HeaderPropagationMiddleware.cs

```diff
@@ -0,0 +1,60 @@
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.HeaderPropagation
+{
+    public class HeaderPropagationMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly HeaderPropagationOptions _options;
+        private readonly HeaderPropagationState _state;
+
+        public HeaderPropagationMiddleware(RequestDelegate next, IOptions<HeaderPropagationOptions> options, HeaderPropagationState state)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+            
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            _options = options.Value;
+
+            _state = state ?? throw new ArgumentNullException(nameof(state));
+        }
+
+        public Task Invoke(HttpContext context)
+        {
+            if (context != null)
+            {
+                foreach (var header in _options.Headers)
+                {
+                    if (!context.Request.Headers.TryGetValue(header.InputName, out var values)
```

Comment:
I think this logic would be clearer if you reversed this condition, and didn't nest `if`s. It took me a few passes through to be sure that it was doing what you wanted.

Reply:
What do you think now? Open for suggestions

Reply:
Looks good